### DISCUSSION TITLE
Is there a hook to follow for starring articles?

### DISCUSSION THREAD

**Discussion by huffstler** [FreshRSS/FreshRSS#6878] [Q&A] [answered]
I'm reading through the extension developer documentation, and found the section outlining the hooks that are available for me to use. I'm not seeing any that allow me to see events for when an article is Starred, I don't think?

High level, I want to write a small extension that enables the following workflow:

- Star an article on my phone (using [Read You](https://github.com/Ashinch/ReadYou))
- This extension runs when the 'star' event hits FreshRSS
  - The article is saved to my Pocket account
  - I can then access the article on my Kobo E-Reader for a better reading experience.

Honestly, like, 95% of the work is already done with [this extension](https://github.com/christian-putzke/freshrss-pocket-button). I would just need the logic to apply when a star event comes in, as opposed to a button in FreshRSS UI being clicked. I'm pretty sure `Read You` is communicating using the GReader api, so it might be a better idea to surface events for applicable api actions instead of just starring?

**Comment by Alkarex** [comment]
I could add an event giving you an array of IDs of entries being starred. Would that help?
Edit: You are welcome to try https://github.com/FreshRSS/FreshRSS/pull/6880 (untested)

**Reply by huffstler** [reply]
That would absolutely work, I'll give this a go when I have some time. Thank you, that's some pretty incredible turnaround 👀

**Reply by huffstler** [reply]
Wanted to check back in here and say I'm still committed to looking at this and testing, just haven't had time and won't for another week (probably more like 2). Getting married takes a lot of attention and time 😁

**Reply by Alkarex** [reply]
Congrats! 💒
Now https://github.com/FreshRSS/FreshRSS/pull/6880 is merged, so it can easily be tested in the `edge` branch (rolling release), for instance with the corresponding git or Docker image `freshrss/freshrss:edge`

**Comment by huffstler** [comment]
More conversation about the nitty-gritty of extension development is happening in the discussion I made in [the extensions repository](https://github.com/FreshRSS/Extensions/discussions/258)

**Comment by huffstler** [comment]
Hey sorry to bother, I finally got the time to stumble through making that extension I was talking about. I got it done, and the hook fires correctly... but only when clicking the Star button through the FreshRSS UI. :\ 

I haven't been able to get the `entries_favorite` hook to fire when favoriting from Read You. I'm looking through the source to find where the logic should fire, but it's slow going for now.

**Reply by Alkarex** [reply]
There:
https://github.com/FreshRSS/FreshRSS/blob/e9ce099277bd5cb0257b9535cfaed42d87670b88/app/Models/EntryDAO.php#L365

and for the GReader API:
https://github.com/FreshRSS/FreshRSS/blob/e9ce099277bd5cb0257b9535cfaed42d87670b88/p/api/greader.php#L819

and for the Fever API:
https://github.com/FreshRSS/FreshRSS/blob/e9ce099277bd5cb0257b9535cfaed42d87670b88/p/api/fever.php#L445

**Reply by huffstler** [reply]
I've spent about 6 hours total trying to be able to debug this application at this point. It has not been fruitful.

From what I can tell, the event _should_ be firing when a star comes in from ReadYou, but I've not been able to set up the development environment to a point where I can actually run a debugger. Or even connect ReadYou to the dev container. I believe Windows firewall is preventing me from accessing the container from outside the host. Or even accessing it from within the host in certain situations... It's bonkers.

I'm giving up on doing this on my windows machine and will spin up a test container on my rpi to test instead. (At some later point) As always, vim + debugging with log statements ends up being easier 😆

**Reply by huffstler** [reply]
@Alkarex (or anyone else who can help)

The issue has to be the way I attempt to set the favorite. [This is the function](https://github.com/huffstler/FreshRSS-StarToPocket/blob/master/extension.php#L27) that's called in my extension.php:

```php
public function handleStar(array $starredEntries, bool $isStarred): void {
  Minz_Log::error('HandleStar called');

  $this->registerTranslates();
  foreach ($starredEntries as $entry) {
    if ($isStarred){
      Minz_Log::error("Bool value is: {$isStarred}");

      $url_array = [
        'c' => 'starToPocket',
        'a' => 'add',
        'params' => [
          'id' => $entry,
        ] 
      ];
  
      /**
      * This is PROBABLY not the correct way to do this?
      */
      Minz_Request::forward($url_array);
    }
  }
}
```

The log statements are showing that I'm getting to where I expected to be in the function. 

As far as log output goes, this is what I'm seeing. (filtering through logs using `podman logs -f testrss | grep '\[error\]'`) The first line is giving me pause as well since the star still saves to FreshRSS, but not to Pocket. When I star an article through the web ui, there's no warning and everything works as expected.

```log
[Fri Nov 22 20:27:20.009713 2024] [authz_core:error] [pid 43:tid 43] [client 10.0.2.100:43116] AH01630: client denied by server configuration: /var/www/FreshRSS/p/reader
FreshRSS[43]: [test] [Fri, 22 Nov 2024 20:27:20 -0500] [error] --- Star Action Confirmed
FreshRSS[43]: [test] [Fri, 22 Nov 2024 20:27:20 -0500] [error] --- Inside truth statement
FreshRSS[43]: [test] [Fri, 22 Nov 2024 20:27:20 -0500] [error] --- HandleStar called
FreshRSS[43]: [test] [Fri, 22 Nov 2024 20:27:20 -0500] [error] --- Bool value is: 1
```

What would be the framework-supported way to save a starred article to pocket from the extension.php here?

Edit: 

Just went a little deeper, the `addAction` from the extension controller seems to not be getting called when the event comes in from the app, but does fire when the Favorite button is clicked from the UI. So I'm either calling it incorrectly, OR it's related to the `[authz+_core:error]` from the first line of the logs above. I'm not sure which case it is.

**Reply by Alkarex** [reply]
Quick answer already: regarding the `denied` issue in your logs, please pull the latest `edge` branch, as it has been fixed with
https://github.com/FreshRSS/FreshRSS/pull/7008

**Reply by Alkarex** [reply]
And indeed, the `Minz_Request::forward` cannot work this way, as it will exit the loop. 
You need to call a `curl_exec()`

**Reply by huffstler** [reply]
> And indeed, the `Minz_Request::forward` cannot work this way, as it will exit the loop. You need to call a `curl_exec()`

Do you mean I should be calling curl_exec on the controller for the extension? Or directly calling the Pocket API?

I've been trying to get the former to work, and I've been unsuccessful calling the `StarToPocket::addAction` controller function from the curl call. In fact, trying to do it this way now prevents starring from the UI to making it to Pocket.

**Reply by huffstler** [reply, marked as answer]
New update: 

I went down the path of putting everything in `extension.php` and it now works properly. I can star articles from Read You as well as the Web UI and they will both propagate to Pocket. I don't know exactly what was going wrong with attempting to call the extension controller, but it wasn't working at all. This actually feels like a slightly better solution anyway, as it cuts out the whole Controller completely, reducing the size of the codebase for the extension.

I'm going to test configuring the whole extension from scratch to make sure I didn't screw other processes up when removing the Controller, but if that all works, then I'll consider the work here done. 

Thank you for your time and patience!

**Reply by huffstler** [reply]
And it's done. Couple small changes were needed, but we are live and running. Thanks again for all the help!
